Cloud registry credentials support

sonnysideup opened the pull request:
## Design

- Implements a generic means of fetching cloud registry credentials (e.g. ECR, GCR, ACR) will support for ECR as a first run
- Controller processes new `DynamicCloudCredentials` field on CR and uses that to fetch credentials and add them to the build job as a new mounted secret
- Build job checks if credentials exist in a known location and adds them to _dynamic_ registries
- Username/Password processing details (from auth token et al) is handled by a separate func to create a standard interface for build job

sonnysideup committed:
updates docker/docker dep

should point to the version used by buildkit but our "replace" clause
was stale

sonnysideup committed:
adds support for pulling ecr credentials

uses the aws v2 sdk
adds unit tests
adds duplicate type for mapping docker auth values but this will be
cleaned up prior to merging

sonnysideup committed:
adds generic design of cloud-provider logic

should accommodate all known cloud-providers when implemented

sonnysideup committed:
adds tests for url multiplexer

sonnysideup committed:
updates to existing cloud registry code

adds IsNoRegistryAuthLoaderFound custom error
adds sample code/notes for upcoming gcr support
cleans up comment cruft

sonnysideup committed:
reworks cloud credentials a bit

returning a strict docker type instead of a custom map because (a) the
caller has a handle to the registry and (b) working with that map felt
unnatural in the rest of the code

also adds `DynamicCloudCredentials` field to ContainerImageBuild
resource definition

sonnysideup committed:
reworks docker credentials

actually leverages the docker type AuthConfig so we can unify usage
across production and consumption of docker config json

sonnysideup committed:
adds e2e propagation of dynamic credentials

processed and added as a secret mount into the build job by the
controller, the credentials are then extracted by an external func that
will grant priority to username/password but will also handle auth
tokens (e.g. returned by ECR) and any other future auth sources we need
to handle.

sonnysideup committed:
Merge remote-tracking branch 'origin/master' into cloud-registry-credentials

sonnysideup committed:
fixes vet issues

sonnysideup committed:
comments out unused cloud registry values

golangci-lint doesn't like these and this is fine because they're not
currently being used

steved commented on internal/credentials/docker.go:
```diff
@@ -1,13 +1,56 @@
+	Auths AuthConfigs `json:"auths"`
+}
+
+// ExtractDockerAuth will return a username/password combination from a JSON-representation of a Docker config file.
+// This function will process credentials from any of the possible auth-sources within that configuration.
+func ExtractDockerAuth(input []byte, host string) (string, string, error) {
```
Can we use `docker/cli` for this or does the API not quite fit?

steved replied:
I was thinking something like (haven't tested it)
```
auth := configfile.New().LoadFromReader(input).GetAuthConfig(host)
return auth.Username, auth.Password, ...
```
It would basically handle the decode / `Auth` stuff for you, not the credential helper stuff.

steved commented on internal/cloud/registry/acr/acr.go:
```diff
@@ -0,0 +1,20 @@
+//var urlRegex = regexp.MustCompile(`^acr\.io$`)
+
+func LoadAuths(ctx context.Context, url string) (*types.AuthConfig, error) {
+	return nil, errors.New("GCR is unsupported")
```
nitpick, ACR

steved commented on internal/cloud/registry/ecr/ecr.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+var (
+	urlRegex = regexp.MustCompile(`^\d{12}\.dkr\.ecr\.[a-z0-9-]+\.amazonaws.com$`)
```
https://github.com/awslabs/amazon-ecr-credential-helper/blob/10153a961e2dd9da9f40f6cc284572252b199c58/ecr-login/api/client.go#L34 is a bit more complicated. I wonder if in general the best way to find the "correct" loader is to just try and grab a token.

sonnysideup replied:
It seems like using the regex pattern from the ECR helper is a better option than the one I wrote. That should be an improvement and easy to write tests against that. 

>I wonder if in general the best way to find the "correct" loader is to just try and grab a token.

I guess I don't understand what you mean when you say "just try". Are you suggesting we don't use regex but instead try and perform auth using all loaders until we hit gold? If so, considering the following: (a) blind auth calls can be expensive (e.g. AWS config loading will hang for several secs if there are no valid credentials, not sure about the others) and (b) I'm not sure what the errors returned by the other libraries are (i.e. how do we discern between an error that indicates "not supported" and "supported but failed to auth"?).

steved replied:
Yeah that was the gist. I hadn't considered the performance cost and the potential error handling constraints, thanks.

steved commented on internal/cloud/registry/gcr/gcr.go:
```diff
@@ -0,0 +1,45 @@
+
+//func init() {
+//	registry.DefaultURLMux().RegisterLoader(urlRegex, LoadAuths)
+//}
```
Maybe pull these out into a separate PR for now? Just doesn't seem right to merge the GCR & ACR files.

steved reviewed (approved):
Minor questions, LGTM overall!

sonnysideup committed:
removes gcr/acr registry files

not supported right now, moving these to a separate PR for tracking

sonnysideup committed:
improves ecr url pattern matching and adds tests

sonnysideup committed:
adds aws account id to ecr auth

so we can target the auth data for the ECR repository in question, not
just load auth for the current aws account

sonnysideup committed:
changes shape of data returned from ecr authorizer

pre-processing auth token so that we return username/password in all
cases

sonnysideup committed:
leverages docker/cli for buildjob auth extraction

instead of writing our own logic

sonnysideup committed:
leverages docker cli from within ExtractDockerAuth

they're backwards-compat logic doesn't handle missing hosts and we want
this common behavior shared when getting docker auth from either a
secret or a file within the system

sonnysideup committed:
fixes jobconfig vol bug

adding vols/mounts to the same field that is uses by the external config
results in a growing list of duplicates. furthermore, we cannot "blank
out" that list since it may contain user-specific vols. hence, this adds
new fields that can safely be reset after use instead of creating a more
complex strategy for using the existing vol/mount fields.
